common/src/main/java/gov/cms/ab2d/common/service/CoverageServiceImpl.java

```diff
@@ -84,57 +103,162 @@ private CoverageSearchEvent getLastEvent(CoveragePeriod period) {
 
-        searchEvent.ifPresent(coverageSearchEvent ->
-                coverageRepo.removeAllByCoveragePeriodAndCoverageSearchEvent(period, coverageSearchEvent));
+        CoveragePeriod coveragePeriod = coveragePeriodRepo.getOne(coveragePeriodIds.get(0));
+
+        Map<String, List<Object[]>> coverageMembership = findCoverageMemberships(pageNumber, pageSize, coveragePeriodIds);
+
+        return coverageMembership.entrySet().stream()
+                .map(obj -> summarizeCoverageMembership(coveragePeriod.getContract(), obj))
+                .collect(toList());
+    }
+
+    @Override
+    public List<CoverageSummary> pageCoverage(int pageNumber, int pageSize, Integer... coveragePeriodIds) {
+        return pageCoverage(pageNumber, pageSize, List.of(coveragePeriodIds));
+    }
+
+    /**
+     * Find page of beneficiaries then look up all of their coverage information for a range of identified coverage
+     * periods.
+     */
+    private Map<String, List<Object[]>> findCoverageMemberships(int pageNumber, int pageSize, List<Integer> coveragePeriodIds) {
+        List<CoveragePeriod> coveragePeriods = coveragePeriodRepo.findAllById(coveragePeriodIds);
+
+        // Look up a page of beneficiaries sorted by id
+        Page<String> beneficiaryPage = coverageRepo.findActiveBeneficiaryIds(coveragePeriods, PageRequest.of(pageNumber, pageSize));
+        List<String> beneficiaries = beneficiaryPage.getContent();
+
+        // Look up coverage information for BATCH_SELECT_SIZE beneficiaries at a time
+        // and populate into hash map
+        Map<String, List<Object[]>> coverageMemberships = new HashMap<>(beneficiaries.size());
+        int startIndex = 0;
+        while (startIndex < beneficiaries.size()) {
+            int endIndex = Math.min(startIndex + BATCH_SELECT_SIZE, beneficiaries.size());
+            List<Object[]> rawResults = coverageRepo.findCoverageInformation(coveragePeriodIds, beneficiaries.subList(startIndex, endIndex));
+            Map<String, List<Object[]>> coverages = rawResults.stream()
+                    .collect(groupingBy(raw -> (String) raw[0]));
+            coverageMemberships.putAll(coverages);
+
+            startIndex = endIndex;
+        }
+
+        return coverageMemberships;
+    }
+
+    /**
+     * Summarize the coverage of one beneficiary for
+     */
+    private CoverageSummary summarizeCoverageMembership(Contract contract, Map.Entry<String, List<Object[]>> membershipInfo) {
+
+        String beneficiaryId = membershipInfo.getKey();
+        List<Object[]> membershipMonths = membershipInfo.getValue();
+
+        if (membershipMonths.size() == 1) {
+            LocalDate start = fromRawResults(membershipMonths.get(0));
+            DateRange range = new DateRange(start, start.plusMonths(1));
+            return new CoverageSummary(beneficiaryId, contract, Collections.singletonList(range));
+        } else {
```

Comment:
With a return above this line, don't need the else which allows the rest of the code in the block to be outdented.